main.go

```diff
@@ -1,56 +1,84 @@
 package main
 
 import (
+	"context"
 	"fmt"
-	"log"
-	"net/http"
+	"log/slog"
+	"os"
 	"time"
 
+	"github.com/Hukyl/genesis-kma-school-entry/mail"
+	mailCfg "github.com/Hukyl/genesis-kma-school-entry/mail/config"
 	"github.com/Hukyl/genesis-kma-school-entry/models"
+	modelsCfg "github.com/Hukyl/genesis-kma-school-entry/models/config"
 	"github.com/Hukyl/genesis-kma-school-entry/rate"
 	"github.com/Hukyl/genesis-kma-school-entry/server"
+	serverCfg "github.com/Hukyl/genesis-kma-school-entry/server/config"
 	"github.com/Hukyl/genesis-kma-school-entry/settings"
-	"github.com/Hukyl/genesis-kma-school-entry/utils"
-	"github.com/gin-gonic/gin"
 )
 
-func NotifyUsers() {
-	rate, err := rate.NewNBURateFetcher().FetchRate("USD", "UAH")
+const emailInterval = 24 * time.Hour
+
+func NotifyUsers(ctx context.Context, apiClient server.Client, mc *mail.Client) {
+	db := apiClient.DB
+
+	rate, err := apiClient.RateFetcher.FetchRate("USD", "UAH")
 	if err != nil {
-		log.Println(err)
+		slog.Warn(err.Error())
 		return
 	}
-	db := models.NewDB()
 	var users []models.User
 	db.Connection().Find(&users)
-	log.Printf("Notifying by email %d users\n", len(users))
+	slog.Info(fmt.Sprintf(
+		"Notifying by email %d users\n",
+		len(users),
+	))
 	for _, user := range users {
 		message := fmt.Sprintf("1 USD = %f UAH", rate.Rate)
-		if err := utils.SendEmail(user.Email, message); err != nil {
-			log.Println(err)
+		if err := mc.SendEmail(ctx, user.Email, message); err != nil {
+			slog.Error(err.Error())
 		}
 	}
 }
 
 func main() {
-	settings.InitSettings()
-	models.NewDB().Migrate()
+	err := settings.InitSettings()
+	if err != nil {
+		slog.Error(fmt.Sprintf("Failed to initialize settings: %s", err))
+		panic(err)
+	}
+
+	ctx := context.Background()
+	ctx = context.WithValue(
+		ctx,
+		settings.DebugKey,
+		os.Getenv("DEBUG") == "true",
+	)
+
+	mc := &mail.Client{Config: mailCfg.NewFromEnv()}
+	apiClient := server.Client{
+		Config:      serverCfg.NewFromEnv(),
+		RateFetcher: rate.NewNBURateFetcher(),
+		DB:          models.NewDB(modelsCfg.NewFromEnv()),
+	}
+	ctx = context.WithValue(ctx, settings.APIClientKey, apiClient)
+	apiClient.Engine = server.NewEngine(ctx)
+
+	err = apiClient.DB.Migrate()
+	if err != nil {
+		slog.Error(fmt.Sprintf("Failed to migrate database: %s", err))
```

Comment:
OPTIONAL: I think a better approach would be 
```
slog.Error("Failed to migrate database, slog.Any("error", err))
```

This way log will produce the following output in LogFMT format:
time=2023-08-04T16:56:03.786-04:00 level=ERROR msg="Failed to migrate database" error=<your error>

or following in JSON
{"time":"2023-08-04T16:58:02.939245411-04:00","level":"ERROR","msg":"Failed to migrate database","error":"<error>"}


So, your error would be in a separate field, other than "msg"! 

What are the benefits? Again, It would be easier to query it when your application is deployed somewhere. Text format is good for humans, but structured formats (JSON, logfmt) are good for machines (centralized logging systems, for example, AWS CloudWatch logs explorer). Imagine, if your application produces ton of logs, and for some reason you need to filter them out. What would be easier? Filter them when output is structured (in AWS you can write powerful queries to filter/sort json) and all additional context (user, errors) is in separate field, or when output is not structured and produced in a plan text-messages format?

The purpose of structured logging is not only to add level to your logs but to add _structure_ to your logs.
I'll give an example from the open-source repo which I've previously contributed to.

[Here](https://github.com/GolangUA/telegram-butler/blob/main/internal/handler/callback/handler.go#L77) error is logged as a separate field. With it I can easily query logs I'm interested it using google cloud console

With your approach - this is not possible

<img width="1457" alt="image" src="https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-Hukyl/assets/93580374/c8eef3ee-ad60-45a2-b4eb-f5464245c313">

Reply:
Makes total sense, working with slog for the first time. Thanks!